### PR TITLE
 mamba checkpointing SSU: two-kernel split + ring-buffer cache for checkpointing SSU

### PR DESCRIPTION
**Pull request by ishovkun**
## 📌 Description

Adds two things to the checkpointing-SSU (Mamba-2 MTP-replay) kernel.

**1. Two-kernel split.** An alternative to the single monolithic kernel that wins
at larger batch. A **precompute** kernel produces the conv1d coefficients (the
C·B contraction and the cumulative-A·dt decays); a **persistent, grid-stride
main** kernel does the state replay + output projection, consuming the
precompute's outputs through caller-provided scratch
(`cb_scaled` / `cumAdt_vec` / `cb_old` / `cumAdt_old` — graph-safe, allocated like
`out`). The public API auto-dispatches on `batch × nheads` (monolith below
~1 work-unit/SM, split above); `algorithm="monolith"|"two-kernel"` forces either
path for benchmarking. The split covers 2- and 4-byte state (bf16/fp16/fp32);
**8-bit state (int8/fp8) always runs the monolith** (`checkpointing_ssu_kernel_8bit`)
— its quantize + per-block-scale flow doesn't factor cleanly into precompute→main.

**2. Ring-buffer cache contract (breaking).** The MTP-replay cache tensors change
from the double-buffered `old_x` / `old_B` / `old_dt` / `old_cumAdt` +
`cache_buf_idx` to single-buffered, head-major **ring buffers**:

| tensor       | shape                                   | dtype |
|--------------|-----------------------------------------|-------|
| `x_cache`    | `(state_cache_size, nheads, L, dim)`    | input |
| `B_cache`    | `(state_cache_size, ngroups, L, dstate)`| input |
| `dt_cache`   | `(state_cache_size, nheads, L)`         | f32   |
| `ring_start` | `(state_cache_size,)`                   | i32   |

with `L = max_window + npredicted`. Live tokens are `(ring_start + j) mod L` for
`j ∈ [0, pnat)`; appends land at `(ring_start + pnat + i) mod L`. The **host owns
all ring bookkeeping** (advancing `ring_start` on flush) — the kernels only read
`ring_start` / `prev_num_accepted` and append. No decay is cached: `cumAdt` is a
prefix sum and is not invariant under the ring-start advance a flush performs, so
it is recomputed from the `dt` ring; on the two-kernel path the precompute stages
it into the `cumAdt_old` scratch for the main. Matches the ReplaySSM contract.

## 🔍 Related Issues

Design follows the ReplaySSM cache contract from vLLM PR vllm-project/vllm#48018.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit`.
- [x] I have installed the hooks with `pre-commit install`.
- [x] The hooks (ruff / clang-format / mypy) pass on the changed files.

## 🧪 Tests

- [x] Tests added/updated — `tests/mamba/test_checkpointing_ssu.py` rewritten for
  the ring contract: two-kernel-vs-monolith bit-exactness across write/no-write,
  `d_split`, pipeline-stages, meta-ring refill, and per-state-dtype (f16/f32/mixed)
  cases; ring wraparound; varlen; non-contiguous strides; determinism. All
  validated against the Triton reference (`replay_selective_state_update.py`).
- [x] All tests passing — the two-kernel-vs-monolith parity battery + the broad
  correctness subset pass on B200.

## Reviewer Notes

- **Breaking change.** The cache API changes shape and semantics (double-buffer →
  ring). Callers must switch to `x_cache / B_cache / dt_cache / ring_start` and
  take over `ring_start` advancement on flush. The custom-op `mutates_args` set
  and the wrapper validation are updated to match.
- **8-bit scope.** The two-kernel split is 2/4-byte-state only; 8-bit stays on the
  monolith (see Description).
- **Perf** (B200/B300, mixed-PNAT sweep with conv1d + PDL — the production shape):
  the two-kernel split overtakes the monolith from ~256 work-units up. The ring
  migration is roughly perf-neutral: the monolith is at parity (slightly faster
  from the gather rewrite) and the two-kernel no-write path is at parity; the
  large-batch two-kernel path carries a small floor (~1.5%) from recomputing the
  decay the ring no longer caches, exposed to ~10% at one low-occupancy batch
  (~32) — a precompute head-tiling tune can close that if needed.

## Summary by CodeRabbit

* **New Features**
  * Added ring-buffer-based SSU checkpointing with monolithic and two-kernel execution modes.
  * Added variable-length sequence support, optional causal convolution integration, and improved cache handling.
  * Added Triton replay reference implementations and expanded CUDA/Triton compatibility options.

* **Benchmarks**
  * Added mixed-checkpointing analysis, PNAT sampling, CSV export, plotting, and benchmark collection tools.

* **Bug Fixes**
  * Improved GPU feature detection and stochastic-rounding support across supported hardware.

* **Tests**
  * Expanded coverage for replay modes, ring caches, varlen inputs, quantized states, determinism, and two-kernel execution.

### CODE REVIEW

File: benchmarks/bench_ssu_checkpoint_mixed.py [hunk 1]
```diff
@@ -0,0 +1,888 @@
...
+        sys.exit(
+            f"no stationary eigenvalue near 1 (closest is {eigvals[idx]}).  "
+            "AL dist may not produce an ergodic chain."
+        )
+    pi = np.real(eigvecs[:, idx])
+    # Iterative refinement: start near the eigenvector, then iterate
+    # P^k to clean up any imaginary leakage from the eig solve.
+    pi = np.maximum(pi, 0.0)
+    if pi.sum() == 0:
+        pi = np.ones(n) / n
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply benchmarks/bench_ssu_checkpoint_mixed.py:183]: fixed in [fe502a5](https://github.com/flashinfer-ai/flashinfer/pull/3975/commits/fe502a57734fe764cd8b01ba43670a5e072d4e12)

File: csrc/checkpointing_ssu.cu [hunk 2]
```diff
@@ -504,6 +519,28 @@ void checkpointing_ssu(
     p.rand_seed = static_cast<const int64_t*>(rs.data_ptr());
   }
 
+  // Two-kernel split scratch (presence of cb_scaled selects precompute+main).
+  // Caller-allocated (graph-safe); both/neither of cb_scaled+cumAdt_vec, plus
+  // cb_old for the no-write path — validated in the Python wrapper.
+  if (cb_scaled.has_value()) p.cb_scaled = const_cast<void*>(cb_scaled.value().data_ptr());
+  if (cumAdt_vec.has_value()) p.cumAdt_vec = const_cast<void*>(cumAdt_vec.value().data_ptr());
+  if (cb_old.has_value()) p.cb_old = const_cast<void*>(cb_old.value().data_ptr());
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply csrc/checkpointing_ssu.cu]: fixed in [080685c](https://github.com/flashinfer-ai/flashinfer/pull/3975/commits/080685c30266e845566e0dcb2e59622d18d15f43)

File: benchmarks/bench_checkpointing_ssu.py [hunk 3]
```diff
@@ -1518,15 +1980,57 @@ def close(self):
...
     _out_path = None
     if _args.output != "-":
         _ts = datetime.now().strftime("%Y%m%d_%H%M%S")
-        _fname = f"bench_checkpointing_ssu_{_ts}.txt"
+        _fname = f"bench_checkpointing_ssu_{_spec_tag(_args)}_{_ts}.txt"
         if _args.output is None:
-            _out_path = os.path.expanduser(f"~/nemo_logs/{_fname}")
+            _out_path = os.path.expanduser(
+                f"/home/scratch.ishovkun_gpu/benchmarks/mamba_decode/{_fname}"
+            )
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply benchmarks/bench_checkpointing_ssu.py]: Fixed in [54d24a1](https://github.com/flashinfer-ai/flashinfer/pull/3975/commits/54d24a1dcdbdc0fa2cd428e7cc49e0fe2c1ac5b9)

File: benchmarks/bench_ssu_checkpoint_mixed.py [hunk 4]
```diff
@@ -0,0 +1,888 @@
...
+import sys
+from pathlib import Path
+
+import numpy as np
+import torch
+
+# Import the in-process benchmarking helpers from bench_checkpointing_ssu.
+SCRIPT_DIR = Path(__file__).parent
+# Shared results dir with collect_checkpointing_ssu_runs.py.
+CSV_DIR = Path("/home/scratch.ishovkun_gpu/benchmarks/mamba_decode")
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply benchmarks/bench_ssu_checkpoint_mixed.py]: fixed in [54d24a1](https://github.com/flashinfer-ai/flashinfer/pull/3975/commits/54d24a1dcdbdc0fa2cd428e7cc49e0fe2c1ac5b9)

File: benchmarks/collect_checkpointing_ssu_runs.py [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+
+import argparse
+import csv
+import subprocess
+import sys
+from pathlib import Path
+
+SCRIPT_DIR = Path(__file__).parent
+# RESULTS_DIR = SCRIPT_DIR
+RESULTS_DIR = Path("/home/scratch.ishovkun_gpu/benchmarks/mamba_decode/")
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply benchmarks/collect_checkpointing_ssu_runs.py]: fixed in [54d24a1](https://github.com/flashinfer-ai/flashinfer/pull/3975/commits/54d24a1dcdbdc0fa2cd428e7cc49e0fe2c1ac5b9)

File: benchmarks/bench_checkpointing_ssu.py [hunk 6]
```diff
@@ -462,23 +564,63 @@ def build_kernel_inputs(
...
+    cu_seqlens = None
+    if varlen:
+        total = batch * mtp_len
+        assert dt.stride(0) == mtp_len * dt.stride(1), (
+            f"dt (b,T) dims not row-major packed: stride(0)={dt.stride(0)}, "
+            f"expected {mtp_len * dt.stride(1)} — can't view as (1, total, ...)"
+        )
+        x = x.view(1, total, nheads, head_dim)
+        B = B.view(1, total, ngroups, d_state)
+        C = C.view(1, total, ngroups, d_state)
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply benchmarks/bench_checkpointing_ssu.py]: fixed in [5d16c98](https://github.com/flashinfer-ai/flashinfer/pull/3975/commits/5d16c98234b49b421c79f2633edba921fc6f8735)

File: csrc/checkpointing_ssu_customize_config.jinja [hunk 7]
```diff
@@ -21,11 +21,13 @@ constexpr int DSTATE = {{ dstate }};
...
-// The wrapper computes this from the runtime tensors and selects the
-// matching JIT URI; the launcher reads HEADS_PER_GROUP directly without
-// a runtime dispatch step.
+// (HEADS_PER_GROUP, NGROUPS) specialization per .so (was 7 via `dispatchRatio`).
+// The wrapper computes these from the runtime tensors and selects the matching
+// JIT URI; the launcher reads them directly without a runtime dispatch step.
+// NGROUPS is compile-time so the persistent main's unflatten (NHEADS = NGROUPS *
+// HEADS_PER_GROUP) divides by compile-time constants only — no runtime div/mod.
 constexpr int HEADS_PER_GROUP = {{ heads_per_group }};
+constexpr int NGROUPS = {{ num_groups }};
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply csrc/checkpointing_ssu_customize_config.jinja:30]: fixed in [cd2859c](https://github.com/flashinfer-ai/flashinfer/pull/3975/commits/cd2859ca2e3b04c96d4858e0cdc9ef0941107578)

File: benchmarks/bench_ssu_checkpoint_mixed.py [hunk 8]
```diff
@@ -0,0 +1,892 @@
...
+IMG_DIR = CSV_DIR / "img"
+sys.path.insert(0, str(SCRIPT_DIR))
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply benchmarks/bench_ssu_checkpoint_mixed.py:63]: Fixed in 54d24a1 — `CSV_DIR` now resolves via the `FLASHINFER_SSU_BENCH_OUTDIR` env var with a cwd default (shared with `collect_checkpointing_ssu_runs.py`); no hardcoded path remains.

File: csrc/checkpointing_ssu.cu [hunk 9]
```diff
@@ -451,6 +442,30 @@ void checkpointing_ssu(
...
+  // 0 = use the launcher's co-residency heuristic; >0 overrides (must divide HEADS_PER_GROUP).
+  {
+    int64_t const hpg = nheads / ngroups;
+    FLASHINFER_CHECK(precompute_heads_per_cta >= 0 &&
+                         (precompute_heads_per_cta == 0 || hpg % precompute_heads_per_cta == 0),
+                     "precompute_heads_per_cta=", precompute_heads_per_cta,
+                     " must be 0 (heuristic) or a positive divisor of HEADS_PER_GROUP "
+                     "(nheads/ngroups=",
+                     hpg, ")");
+  }
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply csrc/checkpointing_ssu.cu]: `main_heads_per_cta` was removed entirely in 27b733f8 (and its forward declaration in bca68678) rather than re-documented — the persistent grid-stride main runs one head per work-unit, so the knob was ineffective (the launcher `(void)`-discarded it). The parameter, its validation, docstring, JIT binding, Python wrapper, and tests were all dropped, so no stale wording remains.

File: flashinfer/mamba/checkpointing_ssu.py [hunk 10]
```diff
@@ -199,11 +225,10 @@ def _checkpointing_ssu_fake(
...
     state: torch.Tensor,
-    old_x: torch.Tensor,
-    old_B: torch.Tensor,
-    old_dt: torch.Tensor,
-    old_cumAdt: torch.Tensor,
-    cache_buf_idx: torch.Tensor,
+    x_cache: torch.Tensor,
+    B_cache: torch.Tensor,
+    dt_cache: torch.Tensor,
+    ring_start: torch.Tensor,
```
Comment (kahyunnam) [inline flashinfer/mamba/checkpointing_ssu.py:227]: is this a breaking API change? can we double check on any downstream affects? This will probably be flagged in API checks for next release

**Review by kahyunnam — APPROVED** [review]
LGTM in general, approving. One nit comment below. 

Also, is it worth testing algorithm="auto" for testing dispatch for two-kernel?

File: flashinfer/mamba/checkpointing_ssu.py (hunk 10 above)
Comment (ishovkun) [reply flashinfer/mamba/checkpointing_ssu.py:227]: Currently no downstream is running this.

**Review by jimmyzho — APPROVED** [review]

**Comment by jimmyzho** [comment]
Merge pending CI tests passing.
